[SPARK-34182][AVRO] Improve error messages when matching Catalyst-to-Avro schemas

xkrogen opened the pull request:
### What changes were proposed in this pull request?
Improve the error messages for incompatibilities between Avro and Catalyst schemas. First, make `AvroSerializer` more similar to `AvroDeserializer` in printing out contextual information such as hierarchical field names. Standardize exception messages in both serializer and deserializer to always include such contextual information, and include a top-level exception which shows the full schemas which were being parsed when the incompatibility was found. Both now print out the hierarchical name for both the Avro and Catalyst fields, since they may be different due to case sensitivity and Avro union handling.

### Why are the changes needed?
The error messages in this type of failure scenario are very lacking in information on the write path (`AvroSerializer`). Below are two examples of messages that provide insufficient information to determine what went wrong (lacking in field names, context about the overall schema structure, etc.).
```
org.apache.spark.sql.avro.IncompatibleSchemaException: Cannot convert Catalyst type IntegerType to Avro type "float".

org.apache.spark.sql.avro.IncompatibleSchemaException: Cannot convert Catalyst type StructType(StructField(bar,IntegerType,true)) to Avro type {"type":"record","name":"test","fields":[{"name":"NOTbar","type":["null","int"],"default":null}]}.
```
The error messages currently existing in `AvroDeserializer` are much better, but still not very internally consistent, and it would be better if they were consistent with the newly added exception messages in `AvroSerializer`.

### Does this PR introduce _any_ user-facing change?
Error messages when there are incompatibilities between Avro and Catalyst schemas will be greatly improved on when writing Avro data using the `avroSchema` option, a little bit improved when reading Avro data, and much more consistent between the two.

Below is an example of a new message. See `AvroSerdeSuite` for more examples.
```
org.apache.spark.sql.avro.IncompatibleSchemaException: Cannot convert Catalyst type STRUCT<`foo`: STRUCT<`bar`: INT>> to Avro type {"type":"record","name":"top","fields":[{"name":"foo","type":"int"}]}
	at org.apache.spark.sql.avro.AvroSerializer.liftedTree1$1(AvroSerializer.scala:83)
...
Caused by: org.apache.spark.sql.avro.IncompatibleSchemaException: Cannot convert Catalyst field 'foo' to Avro field 'foo' because schema is incompatible (sqlType = STRUCT<`bar`: INT>, avroType = "int")
	at org.apache.spark.sql.avro.AvroSerializer.newConverter(AvroSerializer.scala:230)
...
```

### How was this patch tested?
New unit test suite, `AvroSerdeSuite`, was added to test corner cases on `AvroSerializer` and `AvroDeserializer` and verify that the exception messages are as expected. Existing tests in `AvroSuite` also continue to pass, with modifications in places where assertions were made about the exceptions that would be thrown.

xkrogen committed:
[SPARK-34182][AVRO] Update error messages

xkrogen commented:
Note that this was split off from PR 31201 per the [request here](https://github.com/apache/spark/pull/31201#issuecomment-763348112).

cc @gengliangwang @cloud-fan @dongjoon-hyun

xkrogen commented:
I think the main point which may be contentious is having two `IncompatibleSchemaException` nested inside one another, one "top-level" which shows the two full schemas (Avro and Catalyst) being compared, and one specific one thrown from the site of the issue which has details on the exact mismatch. I think it's really helpful for debugging purposes to be able to see the full context -- many times I have gotten support requests from users with schema issues and I have a hard time tracking down exactly what schema they were using (it might not even be the schema they expected they were using, for example if there was a class name conflict with the Avro class they extracted the schema from). If people don't like the nested exception approach, I'm also open to including the full schemas as a message suffix on the individual exceptions that are thrown.

SparkQA commented:
**[Test build #134467 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134467/testReport)** for PR 31333 at commit [`2d32ce6`](https://github.com/apache/spark/commit/2d32ce6e2361bfa3db1b29d5a2589595f35c82fa).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39053/

SparkQA commented:
**[Test build #134467 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134467/testReport)** for PR 31333 at commit [`2d32ce6`](https://github.com/apache/spark/commit/2d32ce6e2361bfa3db1b29d5a2589595f35c82fa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134467/

xkrogen commented:
cc also @viirya @bozhang2820 @lindblombr who I see have some fingerprints in relevant files in case they have any feedback

cloud-fan commented on core/src/main/scala/org/apache/spark/TestUtils.scala:
```diff
@@ -223,24 +224,37 @@ private[spark] object TestUtils {
-  def assertExceptionMsg(exception: Throwable, msg: String, ignoreCase: Boolean = false): Unit = {
-    def contain(msg1: String, msg2: String): Boolean = {
+  def assertExceptionMsg[E <: Throwable : ClassTag](
+    exception: Throwable,
```
nit: 4 spaces indentation for method parameters.

xkrogen replied:
Argh. Try as I might, I can't convince my IDE to use 4 spaces here. Thanks for the catch.

cloud-fan commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:
```diff
@@ -64,30 +65,35 @@ private[sql] class AvroDeserializer(
+    }
+  } catch {
+    case ise: IncompatibleSchemaException => throw new IncompatibleSchemaException(
+      s"Cannot convert Avro type $rootAvroType to Catalyst type $rootCatalystType.", ise)
```
nit `${rootCatalystType.sql}` to be more readable.

xkrogen replied:
Thanks for the tip, this is much better! 👍🏼

cloud-fan commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:
```diff
@@ -99,7 +105,13 @@ private[sql] class AvroDeserializer(
+    val errorPrefix = s"Cannot convert Avro ${toFieldStr(avroPath)} to " +
+        s"Catalyst ${toFieldStr(sqlPath)} because "
+    val incompatibleMsg = errorPrefix +
+        s"schema is incompatible (avroType = $avroType, sqlType = $catalystType)"
```
ditto, call `.sql` for the catalyst type

cloud-fan commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:
```diff
@@ -67,15 +68,20 @@ private[sql] class AvroSerializer(
+      }
+    } catch {
+      case ise: IncompatibleSchemaException => throw new IncompatibleSchemaException(
+        s"Cannot convert Catalyst type $rootCatalystType to Avro type $rootAvroType.", ise)
```
ditto

cloud-fan commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:
```diff
@@ -215,28 +225,40 @@ private[sql] class AvroSerializer(
-          s"Avro type $avroType.")
+      case _ =>
+        throw new IncompatibleSchemaException(errorPrefix +
+          s"schema is incompatible (sqlType = $catalystType, avroType = $avroType)")
```
ditto

xkrogen committed:
Fix indentation and use '.sql' when printing out Catalyst data types.

xkrogen commented:
Updated to use the `.sql` method on Catalyst datatypes instead of the default `toString` to be more human-friendly. Also updated the PR description to reflect the new messages.

cloud-fan commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:
```diff
@@ -93,7 +99,10 @@ private[sql] class AvroSerializer(
 
-  private def newConverter(catalystType: DataType, avroType: Schema): Converter = {
+  private def newConverter(catalystType: DataType, avroType: Schema,
+    catalystPath: Seq[String], avroPath: Seq[String]): Converter = {
```
shall we use a consistent name? either `sqlPath` or `catalystPath` is fine.

xkrogen replied:
I used `sqlPath` in `AvroDeserializer` because `getRecordWriter` was already calling the Catalyst datatype `sqlType` and I wanted to be consistent. But now I see that even `AvroDeserializer#newWriter` is referring to this as `catalystType`, so I will make it consistent throughout. Thanks for the callout.

cloud-fan commented on external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSerdeSuite.scala:
```diff
@@ -0,0 +1,165 @@
+   * on the field assembler using `f`.
+   */
+  private def createNestedAvroSchemaWithFields(
+    nestedRecordFieldName: String,
```
nit: 4 spaces indentation for parameters.

xkrogen replied:
Argh. This time I had the good sense to audit for any more bad instances and found one additional spot. Thank you.

cloud-fan commented on external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSerdeSuite.scala:
```diff
@@ -0,0 +1,165 @@
+   * on the field assembler using `f`.
+   */
+  private def createAvroSchemaWithTopLevelFields(
+    f: SchemaBuilder.FieldAssembler[Schema] => SchemaBuilder.FieldAssembler[Schema]): Schema = {
```
ditto

SparkQA commented:
**[Test build #134794 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134794/testReport)** for PR 31333 at commit [`fd8e58b`](https://github.com/apache/spark/commit/fd8e58b2c2566cd268a075a5fa7bbce233f50b71).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39382/

SparkQA commented:
**[Test build #134794 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134794/testReport)** for PR 31333 at commit [`fd8e58b`](https://github.com/apache/spark/commit/fd8e58b2c2566cd268a075a5fa7bbce233f50b71).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134794/

xkrogen committed:
Minor formatting fixes. 'sqlType' -> 'catalystType' and 'sqlPath' -> 'catalystPath' for consistency. Standardize on referring to 'SQL type' in error messages instead of 'Catalyst type'

xkrogen commented:
Fixed the unit test failure, which was expecting the old exception message.

I standardized the in-code names to be like `catalystPath`/`catalystType` instead of `sqlPath`/`sqlType`, and since we are now printing SQL-style types in the exception messages, I referred to them as "SQL type" instead of "Catalyst type" in the messages. I think this is more user-friendly anyway.

SparkQA commented:
**[Test build #134801 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134801/testReport)** for PR 31333 at commit [`94540cd`](https://github.com/apache/spark/commit/94540cd9e05dac59e11e2d93339262846d38bf49).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39389/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39389/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39389/

SparkQA commented:
**[Test build #134801 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134801/testReport)** for PR 31333 at commit [`94540cd`](https://github.com/apache/spark/commit/94540cd9e05dac59e11e2d93339262846d38bf49).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134801/

cloud-fan commented:
thanks, merging to master!

xkrogen commented:
Thank you for the reviews @cloud-fan ! They were very helpful.
